### ISSUE TITLE
`cargo run`, on a freshly-created cargo project, does not work until after doing a plain `cargo build`.

### ISSUE THREAD

**Issue by bismurphy** [rust-lang/cargo#11367]
### Problem

I'm attempting to replicate the tutorial here: https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html

`cargo new guessing_game` gives `Created binary (application) 'guessing_game' package`
Then
`cd guessing_game`, followed by `cargo run` gives:
```
Compiling guessing_game v0.1.0 (/home/bismurphy/Projects/Rust/guessing_game)
    Finished dev [unoptimized + debuginfo] target(s) in 2.49s
     Running `target/debug/guessing_game`
error: could not execute process `target/debug/guessing_game` (never executed)

Caused by:
  Invalid argument (os error 22)
```
This is odd - it successfully compiles, but won't run. So then I do:
`cd ..`, `rm -r guessing_game`, `cargo new guessing_game`, `cd guessing_game`, `cargo build`, `cargo run`

And at that point, it runs successfully. 

### Steps

`cargo new guessing_game` gives `Created binary (application) 'guessing_game' package`
Then
`cd guessing_game`, followed by `cargo run` gives:
```
Compiling guessing_game v0.1.0 (/home/bismurphy/Projects/Rust/guessing_game)
    Finished dev [unoptimized + debuginfo] target(s) in 2.49s
     Running `target/debug/guessing_game`
error: could not execute process `target/debug/guessing_game` (never executed)

Caused by:
  Invalid argument (os error 22)
```
This is odd - it successfully compiles, but won't run. So then I try again, with an explicit build step added:
`cd ..`, `rm -r guessing_game`, `cargo new guessing_game`, `cd guessing_game`, `cargo build`, `cargo run`

And at that point, it runs successfully. If I do a dedicated compile step prior to running, then all is well. Additionally, if I create the project, then do `cargo run` twice, the first one gives me the error, but the second one works fine. I wonder if there might be some sort of race condition going on, where the compiling step is done, but the file isn't fully written, and the system needs just a tiny bit more time to finish getting the file written to disk before being ready to actually run it.

### Version

```text
cargo 1.65.0 (4bc8f24d3 2022-10-20)
release: 1.65.0
commit-hash: 4bc8f24d3e899462e43621aab981f6383a370365
commit-date: 2022-10-20
host: x86_64-unknown-linux-gnu
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1q)
os: Linux Mint 21 (vanessa) [64-bit]
```

**Comment by ehuss** [comment]
Hm, that's strange. Does the system perhaps have an anti-virus program running or any other kind of background process that may be interfering?

Can you maybe run it under strace to see where the error is coming from? If you are using rustup, you'll need to run something like this to avoid tracing the rustup wrapper:

```bash
strace `rustup which cargo` run
```

**Comment by bismurphy** [comment]
Okay, I think I got that. Not feeling great about what starts at line 252, could you let me know if I'm leaking some sort of secret? https://gist.github.com/bismurphy/6112cc7cfbd6c0c77d3c43afe35a01e6

It doesn't match any of my SSH keys or anything, but I'm not sure where that's coming from - if it's a problem I'd like to know so I can delete. Thanks!

**Comment by ehuss** [comment]
`/usr/lib/ssl/certs/ca-certificates.crt` are the public certificates for the root servers for SSL, I don't think it's anything to be concerned about.

It looks like you are using a samba mount. I would suspect that is the culprit. Can you perhaps use a locally mounted drive?

**Comment by bismurphy** [comment]
Yes, with a local drive it works fine. I make my home directory a samba mount so that I have access to all my personal files no matter what computer I'm using, including if I decide to reinstall my operating system on any of them.

It still seems odd that it would work this way on only a samba instance. I've never had any other programs struggle like this. Is there anything we can do to debug why the difference happens over samba?

**Comment by ehuss** [comment]
Samba is a pretty complicated beast, and it might take a lot of work to narrow down which combination of the environment that is causing the issue. It will likely depend on what the server is running, which version of the samba protocol it is using, the settings on the server, the settings on the client, the version on the client, etc. I don't have any particular insight into why it might be happening for you.

Searching for [samba+einval](https://www.google.com/search?client=firefox-b-1-d&q=%22samba%22+%22einval%22) gives quite a lot of results if one wants to go digging.
